xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/objects/classes/DBListClass.java

```diff
@@ -282,8 +290,10 @@ public String getQuery(XWikiContext context)
                     select.append("doc." + idField);
                 } else {
                     select.append("idprop.value");
-                    fromStatements.add("StringProperty as idprop");
                     whereStatements.add("obj.id=idprop.id.id and idprop.id.name='" + idField + "'");
+                    // Get the from statements according to the type of property
+                    String classType = getPropertyType(classname, idField, context);
+                    fromStatements.add(classType + " as idprop");
```

Comment:
Any reason why you swapped the ``fromStatements`` with the ``whereStatements``. Syntactically there's no problem, but semantically the FROM clause is before the WHERE clause and thus it's easier to read the code if you keep them in order.